Set QP resource defaults

skonto opened the pull request:
Part of the work for https://github.com/knative/serving/issues/13861.


<!-- Please include the 'why' behind your changes if no issue exists -->
## Proposed Changes

* Sets the deployment cm resource values by default.
* Defaults are useful to avoid quota issues and guide users of what is the recommended values to use

**Release Note**

<!-- Enter your extended release note in the below block. If the PR requires
additional action from users switching to the new release, include the string
"action required". If no release note is required, write "NONE". -->

```release-note
A new flag is introduced `queueproxy.resource-defaults` that sets resource requests, limits for Queue Proxy when enabled.
```

skonto commented:
/assign @dprotaso

xtreme-vikram-yadav commented:
lgtm

dprotaso commented:
Should we deprecate the % knob?

ReToCode commented:
I agree with Daves comments, the current requests seem very high, as these are blocking resources that we probably do not use. I still have a todo open to get the real world scenario numbers for various components. This might help to find better defaults in a follow-up PR.

@dprotaso I do not see a way how this is not breaking. Vanilla K8S does not define any defaults, so we don't get any requests & limits today. We cannot not break that if we define something. Can we add it as a breaking change to the release notes?

evankanderson commented:
> @dprotaso I do not see a way how this is not breaking. Vanilla K8S does not define any defaults, so we don't get any requests & limits today. We cannot not break that if we define something. Can we add it as a breaking change to the release notes?

Dave's point was specifically about defining _limits_ where they don't exist today.  The k8s default value for `limit` is "unlimited" unless a value is set, so we're effectively reducing from ♾️  -> some finite value.  This is probably good, particularly for a noncompressible resource like memory, but it could break someone who has a lot of really big requests buffered in-flight.

ReToCode commented:
> Dave's point was specifically about defining limits where they don't exist today.

> Vanilla K8S does not define any defaults, so we don't get any requests & limits today

Exactly. This is why I think it will always be a breaking change. Can we do that when announcing it as breaking change in the release notes?

dprotaso commented:
FWIW - it looks like downstream we're setting our defaults to

```
  queue-sidecar-cpu-request: "25m"
  queue-sidecar-cpu-limit: "1000m"
  queue-sidecar-memory-request: "50Mi"
  queue-sidecar-memory-limit: "200Mi"
```



> Can we do that when announcing it as breaking change in the release notes?

Yeah. I also think we need a way to 'unset' this so we can apply this default to our KinD installations and give operators a way to 'preserve' past behaviour.

skonto commented:
> Yeah. I also think we need a way to 'unset' this so we can apply this default to our KinD installations and give operators a way to 'preserve' past behaviour.

I will add a flag to make migration smoother as with `secure-pod-defaults`. I will add `qp-resource-defaults" if there is no objection. At some point we can just set this to "enabled" by default.

skonto commented on config/core/configmaps/deployment.yaml:
```diff
@@ -57,23 +57,23 @@ data:
 
     # Sets the queue proxy's ephemeral storage request.
-    # If omitted, no value is specified and the system default is used.
+    # If omitted, a default value (currently "512Mi"), is used.
```
I was not sure about that either, there is a interesting history behind the validation of the quota resources: https://github.com/kubernetes/kubernetes/issues/110956. Also ephemeral storage is not validated as reported in the ticket. So I think its safe to remove (I will double check).

skonto replied:
After some experimentation it seems to me that ephemeral storage should be skipped as there is no guarantee that someone will use it if set.

dprotaso commented:
> I will add a flag to make migration smoother as with secure-pod-defaults. I will add qp-resource-defaults if there is no objection. At some point we can just set this to "enabled" by default.

Sounds good

skonto commented on config/core/configmaps/deployment.yaml:
```diff
@@ -57,23 +57,23 @@ data:
 
     # Sets the queue proxy's ephemeral storage limit.
-    # If omitted, no value is specified and the system default is used.
+    # If omitted, a default value (currently "1024Mi"), is used.
```
Config map validation does not allow an empty string as the quantity is parsed: 
```
error: configmaps "config-deployment" could not be patched: admission webhook "config.webhook.serving.knative.dev" denied the request: validation failed: failed to parse "queue-sidecar-memory-limit": quantities must match the regular expression '^([+-]?[0-9.]+)([eEinumkKMGTP]*[-+]?[0-9]*)$'
```

skonto commented:
@dprotaso gentle ping. Status:
- Added the flag for smooth transition, users can choose to disable the flag for a new release that turns it on by default.
- Added some test scenarios
- Updated the release note

dprotaso commented:
/lgtm
/approve

@skonto can you update the release note to say it only applies to cpu and memory?
